src/magic_folder/scanner.py

```diff
@@ -0,0 +1,129 @@
+    #   stoppped, so that it will have no pending work.
+    return Cooperator(
+        scheduler=schedule,
+    )
+
+
+@attr.s
+class ScannerService(Service):
+    """
+    Periodically scan a local Magic Folder for new or updated files
+    """
+
+    _config = attr.ib()
+    _local_snapshot_service = attr.ib()
+    _status = attr.ib()
+    _cooperator = attr.ib()
+    _lock = attr.ib(init=False, factory=DeferredLock)
+
+    @classmethod
+    def from_config(cls, clock, folder_config, local_snapshot_service, status):
+        return cls(
+            config=folder_config,
+            local_snapshot_service=local_snapshot_service,
+            status=status,
+            cooperator=_create_cooperator(clock),
+        )
+
+    def __attrs_post_init__(self):
+        super(ScannerService, self).__init__()
+
+    def stopService(self):
+        super(ScannerService, self).stopService()
+        self._cooperator.stop()
+
+    def scan_once(self):
+        """
+        Perform a scan for new files.
+        """
+        return self._scan()
+
+    @exclusively
+    @inline_callbacks
+    def _scan(self):
+        """
+        Perform a scan for new files, and wait for all the snapshots to be
+        complete.
+        """
+        # TODO: Do we always want to wait for all the files to be snapshotted?
+        # If we don't wait for snapshotting to complete, then we probably want
+        # to coalesce multiple outstanding snapshot requests for the same file,
+        # or otherwise ensure that we don't end up snapshotting an unchanged
+        # file.
+        results = []
+
+        def process(path):
+            d = self._local_snapshot_service.add_file(path)
+            d.addErrback(write_failure)
+            results.append(d)
+
+        with start_action(action_type="scanner:find-updates"):
+            yield find_updated_files(self._cooperator, self._config, process)
+            yield gatherResults(results)
+        # XXX update/use IStatus to report scan start/end
+
+
+def find_updated_files(cooperator, folder_config, on_new_file):
+    """
+    :param Cooperator cooperator: The cooperator to use to control yielding to
+        the reactor.
+    :param MagicFolderConfig folder_config: the folder for which we
+        are scanning
+
+    :param callable on_new_file: a 1-argument callable. This function
+        will be invoked for each updated / new file we find. The
+        argument will be a FilePath of the updated/new file.
+
+    :returns: the scan duration, in seconds
+    """
+    # XXX we don't handle deletes
+    def _process():
+        for path in folder_config.magic_path.walk():
+            if path.isdir():
+                continue
+            relpath = "/".join(path.segmentsFrom(folder_config.magic_path))
+            name = path2magic(relpath)
+            try:
+                snapshot_state = folder_config.get_currentsnapshot_pathstate(name)
```

Comment:
`get_currentsnapshot_pathstate` (and `get_pathinfo`) are both what r0ml might call "potato" interfaces.  I don't think there's much to do about it right now but when we start worrying about poor performance I suspect a fruitful area of investigation will be to vectorize these so we can ask the database for snapshot pathstat for _all_ (or at least N) of the interesting paths at once (dunno if there is something similar we can do for `get_pathinfo` or not ... maybe depends on the platform).

Reply:
If and when we do, I think we'll need to be careful that we get both bits of info we are going to compare without yielding. That would probably mean unrolling this loop in the cooperator, or something.